crates/boundless-market/src/nonce_layer.rs

```diff
@@ -0,0 +1,136 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use alloy::{
+    network::{Ethereum, EthereumWallet, NetworkWallet, TransactionBuilder},
+    primitives::Address,
+    providers::{
+        fillers::{FillProvider, TxFiller},
+        PendingTransactionBuilder, Provider, RootProvider, SendableTx, WalletProvider,
+    },
+    rpc::types::TransactionRequest,
+    transports::{RpcError, TransportResult},
+};
+use std::collections::HashMap;
+use std::sync::Arc;
+use tokio::sync::{Mutex, Semaphore};
+
+/// A provider that manages nonces per account using semaphores.
+///
+/// This provider exists to avoid nonce collisions when submitting transactions concurrently.
+/// It does so by holding a semaphore permit between fetching the pending nonce of the signer until
+/// the transaction is sent.
+#[derive(Clone, Debug)]
+pub struct NonceProvider<F, P>
+where
+    F: TxFiller<Ethereum>,
+    P: Provider<Ethereum> + Send + Sync,
+{
+    inner: Arc<FillProvider<F, P, Ethereum>>,
+    wallet: EthereumWallet,
+    account_semaphores: Arc<Mutex<HashMap<Address, Arc<Semaphore>>>>,
+}
+
+impl<F, P> NonceProvider<F, P>
+where
+    F: TxFiller<Ethereum>,
+    P: Provider<Ethereum> + Send + Sync,
+{
+    /// Construct a new provider with the inner filler and wallet.
+    pub fn new(inner: FillProvider<F, P, Ethereum>, wallet: EthereumWallet) -> Self {
+        Self {
+            inner: Arc::new(inner),
+            wallet,
+            account_semaphores: Arc::new(Mutex::new(HashMap::new())),
+        }
+    }
+
+    /// Get or create a semaphore for the given account address.
+    async fn get_account_semaphore(&self, address: Address) -> Arc<Semaphore> {
+        let mut semaphores = self.account_semaphores.lock().await;
+        semaphores.entry(address).or_insert_with(|| Arc::new(Semaphore::new(1))).clone()
+    }
+}
+
+#[async_trait::async_trait]
+impl<F, P> Provider<Ethereum> for NonceProvider<F, P>
+where
+    F: TxFiller<Ethereum>,
+    P: Provider<Ethereum> + Send + Sync + std::fmt::Debug,
+{
+    fn root(&self) -> &RootProvider<Ethereum> {
+        self.inner.root()
+    }
+
+    async fn send_transaction(
+        &self,
+        mut request: TransactionRequest,
+    ) -> TransportResult<PendingTransactionBuilder<Ethereum>> {
+        let from_address = if let Some(from) = request.from {
+            from
+        } else {
+            <EthereumWallet as NetworkWallet<Ethereum>>::default_signer_address(&self.wallet)
+        };
+        request.set_from(from_address);
+
+        // Get semaphore for this account and acquire permit
+        let semaphore = self.get_account_semaphore(from_address).await;
+        let _permit = semaphore.acquire().await.unwrap();
+
+        // Fetch the pending nonce if not already set
+        if request.nonce.is_none() {
```

Comment:
@austinabell are we sure that the default filler are not going to fill this nonce up and that happen without semaphore control?

Reply:
nvm, the fill happens later, all good!

Reply:
yeah this is only set if the caller explicitly sets a nonce, but we don't do that here. 

Thinking about this though, we should probably disable recommended fillers to at least simplify the flow I think?